ADR-0002 - layer deployments

roleyfoley opened the pull request:
This ADR aims to outline the possible options that we have to support deployments of resources at the layer level ( Tenant, Account, Product, Environment ) etc. 

The need to be able to define shared deployments which could have complex requirements to our component deployments has  brought this issue up and so an ADR has been started to discuss how we approach it

roleyfoley committed:
feat: 002 - layer deployments

roleyfoley committed:
fix: update ADR log

roleyfoley commented:
@hamlet-io/engine-maintainers Whats the feeling here? are we happy with this approach?

RossMurr4y commented:
I think it's a workable solution.

Do you envision any existing Components becoming Foundations?

roleyfoley commented:
The current AWS account deployments would be the primary candidate. None of the existing components would qualify I don't think

RossMurr4y commented:
Does the name "Foundations" suggest a certain level of prerequisite, or underlying importance to the product that is not necessarily the case for layer-level components? If something is a layer-level component due to a providers limitation on a given resource, then it isn't necessarily either.

There also isn't necessarily agreement between providers on resource limitations either. If AWS says "one of resource X per account" but the makeup of Services/Resources for the same component in Azure has no such limitations, then it seems like we'll have to make a per-component decision about which one to enforce, or choose to deliberately support the components differently per-provider (a "Foundation" in AWS and component in Azure, in this example).

roleyfoley commented:
That was the intention of the name. The account level deployments are currently required to make things work. I.e. if you don't have a registry then we can't deploy code. 

I don't think these would be shared level components and would instead be implemented as required by providers. So effectively it would be the foundations for using a given provider with hamlet

roleyfoley committed:
feat: add layer occurrence process and descision

Adds a new option of making the existing occurrence build process
based to support working on layers. This is now the preferred option due
to its potential options and future extensibility

roleyfoley committed:
fix: extend description

ml019 reviewed (approved):
Overall the refactoring pain of this option is worth it given the consistency it introduces.
